apps/consulting/posts/scaling-python-banking-edition.md

```diff
@@ -0,0 +1,677 @@
+        delayed_df = summed_df.to_delayed(optimize_graph=True)
+    info_delayed = info(delayed_df)
+    return info_delayed
+```
+
+<img
+src="/posts/scaling-python-banking-edition/dask-bag.png"
+alt="A picture of a jupyter notebook input cell, next to an image of the cell contents equivilent dask graph being run. The input cell has the code for a dask bag data frame summation wit resource annotations."
+/>
+
+**_!!!! (We need to walk the reader through the above figure) !!!!_**
+
+On top of Dask Bag, we also started utilizing a tool called 'resource
+annotations.' These allowed us to take a much more
+hands-off approach to the scheduler. Instead of manually managing how much work
+we are submitting at a time, we can submit it all and let the scheduler do the
+heavy lifting of resource management. This automatically insures that the workers will
+not get work that exceeds the resources that they have available.
+
+To use resource annotations, first you need to define the relevant resource
+limits when you set up your cluster.
+
+```python
+with dask.config.set({'distributed.worker.resources.MEMORY': 2000}):
+    cluster = LocalCluster()
+client = Client(cluster)
+```
+
+As you can see in the above example, I used the context manager
+`dask.config.set` to set the `MEMORY` resources of each worker
+to 2000 **_WHAT UNITS?_**. The resource name and the value can be set arbitrarily. For
+example, I could have set the resource to be `WIDGET` instead of
+`MEMORY`, and it would work the same as long as your naming is consistent
+between your cluster configuration and your annotations. Next, you
+annotate your Dask **_Dask Bag?_** objects.
+
+```python
+import dask.bag as db
+
+part_size=10
+with dask.annotate(resources={'MEMORY': 200*part_size}):
+    b = db.from_sequence(filepaths, partition_size=part_size)
+    dfs = b.map(load_dataframe)
+    summed_df = dfs.sum()
+```
+
+In the above code, I have annotated the blocks of Dask objects with
+the max amount of memory I expect them to need. Now, when I run this,
+the scheduler knows not to send off a block of work to a worker until
+there are the necessary resources available on a worker.
+
+## Orchestration & Productionisation
+
+With the computational machinery defined, we then needed a tool to orchestrate all the work for us.
+In particular, we needed something that we could submit the work to, which would then handle
+making sure all the pieces were put together and run correctly. We
+tried two separate tools for this, Prefect and Argo Workflows.
+
+### Prefect
+
+We initially thought that Prefect would be a good fit for us. Prefect
+is a Python-native package, which was a bonus. Also, tasks in Prefect
+are just Python functions wrapped in a workflow definition with some flow control
+logic. It includes a server that can be used to run flows in response
+to triggers and schedulers and to manage logging, and it provides a
+visualization dashboard. It also integrates nicely with Dask to run
```

Comment:
"It also integrates nicely with Dask to run each Dask on a Dask cluster."

I think something went wrong with this sentence, and I can't figure out what the intent was. Can you suggest a revision, or describe in more detail what you were trying to say?

Reply:
Yeah, its supposed to be "It also integrates nicely with Dask to run each task on a Dask cluster."

Maybe we can add a little more there so it's "It also integrates nicely with Dask to run each task on a Dask cluster with little edits to the code." or something like that?

Reply:
Fixed this. Let me know what you think.